grpc/src/client/name_resolution/dns/mod.rs

```diff
@@ -0,0 +1,397 @@
+//! in grpc.
+
+use std::{
+    net::{IpAddr, SocketAddr},
+    sync::{
+        atomic::{AtomicU64, Ordering},
+        Arc, Mutex,
+    },
+    time::{Duration, SystemTime},
+};
+
+use tokio::sync::Notify;
+use url::Host;
+
+use crate::{
+    client::name_resolution::{Address, NopResolver, ResolverUpdate, TCP_IP_NETWORK_TYPE},
+    rt,
+};
+
+use super::{
+    backoff::{BackoffConfig, ExponentialBackoff, DEFAULT_EXPONENTIAL_CONFIG},
+    global_registry, Endpoint, Resolver, ResolverBuilder,
+};
+
+#[cfg(test)]
+mod test;
+
+const DEFAULT_PORT: u16 = 443;
+const DEFAULT_DNS_PORT: u16 = 53;
+
+/// This specifies the maximum duration for a DNS resolution request.
+/// If the timeout expires before a response is received, the request will be
+/// canceled.
+///
+/// It is recommended to set this value at application startup. Avoid modifying
+/// this variable after initialization.
+static RESOLVING_TIMEOUT_MS: AtomicU64 = AtomicU64::new(30_000); // 30 seconds
+
+/// This is the minimum interval at which re-resolutions are allowed. This helps
+/// to prevent excessive re-resolution.
+static MIN_RESOLUTION_INTERVAL_MS: AtomicU64 = AtomicU64::new(30_000); // 30 seconds
+
+fn get_resolving_timeout() -> Duration {
+    Duration::from_millis(RESOLVING_TIMEOUT_MS.load(Ordering::Relaxed))
+}
+
+/// Sets the maximum duration for DNS resolution requests.
+///
+/// This function affects the global timeout used by all channels using the DNS
+/// name resolver scheme.
+///
+/// It must be called only at application startup, before any gRPC calls are
+/// made.
+///
+/// The default value is 30 seconds. Setting the timeout too low may result in
+/// premature timeouts during resolution, while setting it too high may lead to
+/// unnecessary delays in service discovery. Choose a value appropriate for your
+/// specific needs and network environment.
+pub fn set_resolving_timeout(duration: Duration) {
+    RESOLVING_TIMEOUT_MS.store(duration.as_millis() as u64, Ordering::Relaxed);
+}
+
+fn get_min_resolution_interval() -> Duration {
+    Duration::from_millis(MIN_RESOLUTION_INTERVAL_MS.load(Ordering::Relaxed))
+}
+
+/// Sets the default minimum interval at which DNS re-resolutions are allowed.
+/// This helps to prevent excessive re-resolution.
+///
+/// It must be called only at application startup, before any gRPC calls are
+/// made.
+pub fn set_min_resolution_interval(duration: Duration) {
+    MIN_RESOLUTION_INTERVAL_MS.store(duration.as_millis() as u64, Ordering::Relaxed);
+}
+
+pub fn reg() {
+    global_registry().add_builder(Box::new(Builder {}));
+}
+
+struct Builder {}
+
+struct DnsOptions {
+    min_resolution_interval: Duration,
+    resolving_timeout: Duration,
+    backoff_config: BackoffConfig,
+}
+
+impl DnsResolver {
+    fn new(
+        target: &super::Target,
+        options: super::ResolverOptions,
+        dns_opts: DnsOptions,
+    ) -> Box<dyn Resolver + 'static> {
```

Comment:
Curious. Would changing this method to return `Self`, and having the `build` method put it in a box be better? Because I would assume the `build` method is the only caller for this, and I feel it is cleaner for a `new` method to return `Self`. I could be wrong.